p2p/nodestate/nodestate.go

```diff
@@ -646,53 +659,104 @@ func (ns *NodeStateMachine) SetState(n *enode.Node, setFlags, resetFlags Flags,
 			node.dirty = true
 		}
 	}
-	ns.lock.Unlock()
-	// call state update subscription callbacks without holding the mutex
-	for _, sub := range ns.stateSubs {
-		if changed&sub.mask != 0 {
-			sub.callback(n, Flags{mask: oldState & sub.mask, setup: ns.setup}, Flags{mask: newState & sub.mask, setup: ns.setup})
+	callback := func() {
+		for _, sub := range ns.stateSubs {
+			if changed&sub.mask != 0 {
+				sub.callback(n, Flags{mask: oldState & sub.mask, setup: ns.setup}, Flags{mask: newState & sub.mask, setup: ns.setup})
+			}
 		}
-	}
-	if newState == 0 {
-		// call field subscriptions for discarded fields
-		for i, v := range node.fields {
-			if v != nil {
-				f := ns.fields[i]
-				if len(f.subs) > 0 {
-					for _, cb := range f.subs {
-						cb(n, Flags{setup: ns.setup}, v, nil)
+		if newState == 0 {
+			// call field subscriptions for discarded fields
+			for i, v := range node.fields {
+				if v != nil {
+					if debugPrints {
+						fmt.Println("discardField", n.ID(), ns.setup.fields[i].name, v)
+					}
+					f := ns.fields[i]
+					if len(f.subs) > 0 {
+						for _, cb := range f.subs {
+							cb(n, Flags{setup: ns.setup}, v, nil)
+						}
 					}
 				}
 			}
 		}
 	}
+	callNow := node.pendingCallbacks == nil
+	node.pendingCallbacks = append(node.pendingCallbacks, callback)
+	ns.callbackCount++
+	list := node.pendingCallbacks
+	ns.lock.Unlock()
+	// call state update subscription callbacks without holding the mutex
+	if callNow {
+		ns.processCallbacks(node, list)
+	}
```

Comment:
I think it will make this tool even harder to use.

Originally we hold the assumption that after we set the flags or fields, all relevant subscriptions will be invoked. But now it's async.

So for example, If I set field f as X, originally I can expect all subscriptions to be called, but now I need additional check. Like in the clientpool, if we can't initialize the balance instance immediately, it may be kicked out.

Reply:
Now it's fixed in another PR which this PR is based on.